### PR TITLE
Implement hover for application.properties keys

### LINKED ISSUE #6: Hover support for application.properties
Show documentation when a key from application.properties is hovered.

Add test in QuarkusAssert

### PR DESCRIPTION
**Pull request by xorye**
Fixes #6 

Example:
image

I did not write tests for this feature yet.
I also did not write an extension/participant for hover.

I might be wrong, but I think there are some small problems with [`PropertiesScannerUtils.java`](https://github.com/xorye/quarkus-ls/blob/3a32cc72b7300f91dbdebad7b3a311ce9915d3bb/quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/utils/PropertiesScannerUtils.java). More specifically [here](https://github.com/xorye/quarkus-ls/blob/3a32cc72b7300f91dbdebad7b3a311ce9915d3bb/quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/utils/PropertiesScannerUtils.java#L135) and [here](https://github.com/xorye/quarkus-ls/blob/3a32cc72b7300f91dbdebad7b3a311ce9915d3bb/quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/utils/PropertiesScannerUtils.java#L141).

For the first one, I think the for loop condition should be `i < text.length()` and for the second one, I think we need to change `startLineOffset`. I can fix these if needed.

Signed-off-by: David Kwon <dakwon@redhat.com>

### CODE REVIEW

**Comment by angelozerr** [comment]
In short, great PR!

> I did not write tests for this feature yet.

Please add test like we have for completion. I suggets you that you see lsp4xml (XMLAssert) to show hover assert and adapt it for Quarkus.

> I also did not write an extension/participant for hover.

No need extension here. LSP4XML provides extension/participant to manage custom XML hover. We don't need that in application.properties.

> I might be wrong, but I think there are some small problems with PropertiesScannerUtils.java. 

> For the first one, I think the for loop condition should be i < text.length() and for the second one, I > think we need to change startLineOffset. I can fix these if needed.

To be honnest with you I have written this code quickly. If there are some troubles please fix it.

See my review comments.

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/ls/QuarkusTextDocumentService.java [hunk 1]
```diff
@@ -108,6 +111,21 @@ public void didSave(DidSaveTextDocumentParams params) {
 		});
 	}
 
+	@Override
+	public CompletableFuture<Hover> hover(TextDocumentPositionParams params) {
+
+		QuarkusProjectInfoParams projectInfoParams = createProjectInfoParams(params.getTextDocument(), null);
+		return projectInfoCache.getQuarkusProjectInfo(projectInfoParams).thenApplyAsync(projectInfo -> {
+			if (!projectInfo.isQuarkusProject()) {
+				return new Hover();
```
Comment (angelozerr) [inline quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/ls/QuarkusTextDocumentService.java]: return null

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusHover.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+import com.redhat.quarkus.utils.DocumentationUtils;
+import com.redhat.quarkus.utils.PropertiesScannerUtils;
+import com.redhat.quarkus.utils.PropertiesScannerUtils.PropertiesToken;
+
+import org.eclipse.lsp4j.Hover;
+import org.eclipse.lsp4j.MarkupContent;
+import org.eclipse.lsp4j.MarkupKind;
+import org.eclipse.lsp4j.Position;
+
+class QuarkusHover {
```
Comment (angelozerr) [inline quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusHover.java]: Add javdoc

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusHover.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+	public Hover doHover(TextDocument document, Position position, QuarkusProjectInfo projectInfo,
+			QuarkusHoverSettings hoverSettings) {
+		Hover hover = new Hover();
+		String text = document.getText();
+		int offset;
+		try {
+			offset = document.offsetAt(position);
+		} catch (BadLocationException e) {
+			LOGGER.log(Level.SEVERE, "In QuarkusHover, position error", e);
+			return hover;
```
Comment (angelozerr) [inline quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusHover.java]: return null

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusHover.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+	 * 
+	 * @param document      the document
+	 * @param position      the hover position
+	 * @param projectInfo   the Quarkus project information
+	 * @param hoverSettings the hover settings
+	 * @return Hover object for the currently hovered token
+	 */
+	public Hover doHover(TextDocument document, Position position, QuarkusProjectInfo projectInfo,
+			QuarkusHoverSettings hoverSettings) {
+		Hover hover = new Hover();
```
Comment (angelozerr) [inline quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusHover.java]: create hover instance after the } catch (BadLocationException e) {

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusHover.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+	/**
+	 * Collect the documentation for property key represented by token
+	 * 
+	 * @param document      the document
+	 * @param token         the token representing the property key being hovered
+	 * @param projectInfo   the Quarkus project information
+	 * @param hoverSettings the hover settings
+	 * @param hover         the hover object to fill with documentation
+	 */
+	private static void collectPropertyKeyDocumentation(TextDocument document, PropertiesToken token, QuarkusProjectInfo projectInfo,
```
Comment (angelozerr) [inline quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusHover.java]: rename `collectPropertyKeyDocumentation` by `getPropertyKeyHover`

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusHover.java [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+				break;
+			case VALUE:
+				// no hover documentation
+				break;
+			}
+		return hover;
+	}
+
+	/**
+	 * Collect the documentation for property key represented by token
```
Comment (angelozerr) [inline quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusHover.java]: Returns the documentation hover for property key represented by token

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusHover.java [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+	 * 
+	 * @param document      the document
+	 * @param token         the token representing the property key being hovered
+	 * @param projectInfo   the Quarkus project information
+	 * @param hoverSettings the hover settings
+	 * @param hover         the hover object to fill with documentation
+	 */
```
Comment (angelozerr) [inline quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusHover.java]: @return the documentation hover for property key represented by token

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusLanguageService.java [hunk 8]
```diff
@@ -26,13 +28,20 @@
...
 		this.completions = new QuarkusCompletions();
+		this.hover = new QuarkusHover();
 	}
 
 	public CompletionList doComplete(TextDocument document, Position position, QuarkusProjectInfo projectInfo,
 			QuarkusCompletionSettings completionSettings, CancelChecker cancelChecker) {
 		return completions.doComplete(document, position, projectInfo, completionSettings, cancelChecker);
 	}
+
+	public Hover doHover(TextDocument document, Position position, QuarkusProjectInfo projectInfo,
```
Comment (angelozerr) [inline quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusLanguageService.java]: Add javadoc

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/settings/QuarkusHoverSettings.java [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+	}
+
+	/**
+	 * Returns <code>true</code> if the client support the given documentation
+	 * format and <code>false</code> otherwise.
+	 * 
+	 * @return <code>true</code> if the client support the given documentation
+	 *         format and <code>false</code> otherwise.
+	 */
+	public boolean isDocumentationFormatSupported(String documentationFormat) {
```
Comment (angelozerr) [inline quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/settings/QuarkusHoverSettings.java]: rename with `isContentFormatSupported(String contentFormat)`

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusHover.java (hunk 5 above)
Comment (xorye) [reply quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusHover.java]: Should this method still be a void method? Or should it return Hover instead?

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusHover.java [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+
+		String keyName = null;
+		try {
+			int keyLength = (token.getEnd() - token.getStart()) + 1;
+			Position positionStart = document.positionAt(token.getStart());
+			String line = document.lineText(positionStart.getLine());
+			keyName = line.substring(0, keyLength).trim();
+		} catch (BadLocationException e) {
+			LOGGER.log(Level.SEVERE, "In QuarkusHover, position error", e);
+		}
```
Comment (angelozerr) [inline quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusHover.java]: return null in this error case

- [ ]

File: quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusHover.java (hunk 5 above)
Comment (angelozerr) [reply quarkus.ls/com.redhat.quarkus.ls/src/main/java/com/redhat/quarkus/services/QuarkusHover.java]: yes I think it should be better to have 

private static Hover getPropertyKeyHover(TextDocument document, PropertiesToken token, QuarkusProjectInfo projectInfo) {

}

**Comment by fbricon** [comment]
Was doing some quick testing and I ran into
```
Error - 6:22:56 PM] Request textDocument/hover failed.
  Message: Internal error.
  Code: -32603 
java.util.concurrent.CompletionException: java.lang.NullPointerException
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314)
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:319)
	at java.base/java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:718)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506)
	at java.base/java.util.concurrent.CompletableFuture.postFire(CompletableFuture.java:610)
	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:649)
	at java.base/java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:479)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)
Caused by: java.lang.NullPointerException
	at org.eclipse.lsp4j.adapters.HoverTypeAdapter.writeContents(HoverTypeAdapter.java:82)
	at org.eclipse.lsp4j.adapters.HoverTypeAdapter.write(HoverTypeAdapter.java:142)
	at org.eclipse.lsp4j.adapters.HoverTypeAdapter.write(HoverTypeAdapter.java:38)
	at com.google.gson.TypeAdapter$1.write(TypeAdapter.java:191)
	at com.google.gson.Gson.toJson(Gson.java:669)
	at org.eclipse.lsp4j.jsonrpc.json.adapters.MessageTypeAdapter.write(MessageTypeAdapter.java:412)
	at org.eclipse.lsp4j.jsonrpc.json.adapters.MessageTypeAdapter.write(MessageTypeAdapter.java:55)
	at com.google.gson.Gson.toJson(Gson.java:669)
	at com.google.gson.Gson.toJson(Gson.java:648)
	at org.eclipse.lsp4j.jsonrpc.json.MessageJsonHandler.serialize(MessageJsonHandler.java:145)
	at org.eclipse.lsp4j.jsonrpc.json.MessageJsonHandler.serialize(MessageJsonHandler.java:140)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageConsumer.consume(StreamMessageConsumer.java:59)
	at com.redhat.quarkus.ls.commons.ParentProcessWatcher.lambda$apply$0(ParentProcessWatcher.java:140)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.lambda$handleRequest$1(RemoteEndpoint.java:281)
	at java.base/java.util.concurrent.CompletableFuture$UniAccept.tryFire(CompletableFuture.java:714)
	... 9 more
```

No sure how

**Comment by fbricon** [comment]
That was when hovering over values

**Comment by xorye** [comment]
@fbricon the new changes will get rid of this NPE and will show this instead:

```
[Trace - 12:29:50 p.m.] Received response 'textDocument/hover - (5)' in 2ms.
No result returned.
```

**Comment by xorye** [comment]
I have made the changes and added tests.

One question, is it normal for hover documentation to show up if I hover over the equals sign? 
Like this:
image

This happens because currently, everything before and including the equals sign is considered the key.

**Comment by angelozerr** [comment]
> One question, is it normal for hover documentation to show up if I hover over the equals sign?
Like this:

No, but it's not very annoying. If you can fix it easily do it, otherwise we keep like this.

**Pull request merged**